v4/export/config.go

```diff
@@ -117,6 +129,53 @@ func (conf *Config) GetDSN(db string) string {
 	return dsn
 }
 
+func (config *Config) ExternalStorage(ctx context.Context) storage.ExternalStorage {
+	config.externalStorageOnce.Do(func() {
+		s, err := config.createExternalStorage(ctx)
+		if err != nil {
+			panic(err)
+		}
+		config.externalStorage = s
+	})
+	return config.externalStorage
+}
+
+func (config *Config) createExternalStorage(ctx context.Context) (storage.ExternalStorage, error) {
+	path := config.OutputDirPath
+	var b *backup.StorageBackend
+	var err error
+	// TODO support other filesystems
+	options := &config.BackendOptions
+	// used for testing
+	endpoint := os.Getenv("DUMPLING_S3_ENDPOINT")
+	if endpoint != "" {
+		options.S3.Endpoint = endpoint
+		options.S3.ForcePathStyle = true
+		options.GCS.Endpoint = endpoint
+	}
```

Comment:
I think we can use [failpoint](https://github.com/pingcap/failpoint) here like we did in https://github.com/pingcap/dumpling/blob/48b61dc62c7e27c99b81d939e4614f07e3f3a9d2/v4/export/dump.go#L194 and https://github.com/pingcap/dumpling/blob/48b61dc62c7e27c99b81d939e4614f07e3f3a9d2/tests/consistency/run.sh#L21, but this way is also good.

Reply:
failpoint looks neat, I'll make it so!